pyqg/model.py

```diff
@@ -371,6 +388,129 @@ def stability_analysis(self,bottom_friction=False):
+            multiple). Must all be attributes of the model with shape
+            :code:`(nz, ny, nx)`.
+        coarsening_function : function
+            The function used to coarsen high-resolution variables when
+            computing forcing. Defaults to spectral truncation followed by an
+            application of the low-res model's numerical filter.
+        spectral_filter : function
+            The function used to filter variables during coarsening with the
+            default spectral truncation :code:`coarsening_function`. Defaults
+            to the numerical filter already used in the low-res model for
+            small-scale dissipation.
+
+        Returns
+        -------
+        forcing : array or list of arrays
+            The subgrid forcing associated with the given quantities and
+            coarsening scheme. If there is only one quantity, this is a single
+            array, and otherwise a list of them.
+        """
+
+        assert self.nx == self.ny, "subgrid forcing function assumes nx==ny"
+
+        if lower_resolution is None:
+            lower_resolution = self.nx // 4
+
+        assert lower_resolution < self.nx, "lower_resolution must be lower"
+        assert lower_resolution % 2 == 0, "lower_resolution must be even"
+
+        return_one = False
+
+        if isinstance(quantities, str):
+            quantities = [quantities]
+            return_one = True
+
+        for q in quantities:
+            assert hasattr(self, q), f"unrecognized property {q}"
+            assert getattr(self, q).shape == self.q.shape, f"{q} has wrong shape"
+
+        # Initialize a lower-res model with the same parameters
+        m1 = self
+        m2 = self._reinitialized(nx=lower_resolution, ny=lower_resolution)
+
+        if coarsening_function is None:
+            def coarsening_function(var):
+                # Convert to spectral (if needed)
+                if var.shape != m1.qh.shape: var = m1.fft(var)
+                # Truncate high-frequency indices
+                keep = m2.qh.shape[1]//2
+                trunc = np.hstack((var[:,:keep,:keep+1], var[:,-keep:,:keep+1]))
+                # Apply filter and normalize FFT
+                filtered = trunc * spectral_filter(m2) * (m1.nx / m2.nx)**2
+                # Convert to real
+                return m2.ifft(filtered)
+
+        # Set its state to the coarsened version
+        m2.q = coarsening_function(m1.qh)
+        m2._invert()
+        m2._calc_derived_fields()
+
+        forcings = []
+
+        for quantity in quantities:
+            # Get the corresponding field
+            x1 = getattr(m1, quantity)
+            x2 = getattr(m2, quantity)
+            # Compute the difference in advected quantities
+            forcing = coarsening_function(m1._advect(x1)) - m2.ifft(m2._advect(x2))
+            forcings.append(forcing)
+
+        if return_one:
+            forcings = forcings[0]
+
+        return forcings
```

Comment:
This function could go in a different pull request if that makes reviewing easier.